File: src/apps/kernel/pes/VPE.cc
/*
 * Copyright (C) 2015, Nils Asmussen <nils@os.inf.tu-dresden.de>
 * Economic rights: Technische Universitaet Dresden (Germany)
 *
 * This file is part of M3 (Microkernel-based SysteM for Heterogeneous Manycores).
 *
 * M3 is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 2 as
 * published by the Free Software Foundation.
 *
 * M3 is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * General Public License version 2 for more details.
 */

#include <base/Common.h>
#include <base/log/Kernel.h>
#include <base/util/Math.h>

#include <thread/ThreadManager.h>

#include "pes/PEManager.h"
#include "pes/VPEManager.h"
#include "pes/VPEGroup.h"
#include "pes/VPE.h"
#include "DTU.h"
#include "Platform.h"
#include "SyscallHandler.h"

namespace kernel {

cycles_t VPE::TIME_SLICE = 6000000;

VPE::VPE(m3::String &&prog, peid_t peid, vpeid_t id, uint flags, epid_t sep, epid_t rep,
         capsel_t sgate, VPEGroup *group)
    : SListItem(),
      SlabObject<VPE>(),
      RefCounted(),
      _desc(peid, id),
      _flags(flags | F_INIT | F_NEEDS_INVAL),
      _pid(),
      _state(DEAD),
      _exitcode(),
      _sysc_ep((flags & F_IDLE) ? SyscallHandler::ep(0) : SyscallHandler::alloc_ep()),
      _group(group),
      _services(),
      _pending_fwds(),
      _name(m3::Util::move(prog)),
      _objcaps(id + 1),
      _mapcaps(id + 1),
      _lastsched(),
      _rbufs_size(),
      _dtustate(),
      _upcsgate(*this, m3::DTU::UPCALL_REP, 0),
      _upcqueue(*this),
      _as(Platform::pe(pe()).has_virtmem() ? new AddrSpace(pe(), id, sep, rep, sgate) : nullptr),
      _headers(),
      _rbufcpy(),
      _requires(),
      _argc(),
      _argv(),
      _epaddr() {
    if(_sysc_ep == EP_COUNT)
        PANIC("Too few slots in syscall receive buffers");
    if(group)
        _group->add(this);

    _objcaps.set(0, new VPECapability(&_objcaps, 0, this));
    _objcaps.set(1, new MGateCapability(&_objcaps, 1, pe(), id, 0, MEMCAP_END, m3::KIF::Perm::RWX));
    for(epid_t ep = m3::DTU::FIRST_FREE_EP; ep < EP_COUNT; ++ep) {
        capsel_t sel = 2 + ep - m3::DTU::FIRST_FREE_EP;
        _objcaps.set(sel, new EPCapability(&_objcaps, sel, id, ep));
    }

    if(!Platform::pe(pe()).has_virtmem())
        _rbufcpy = MainMemory::get().allocate(RECVBUF_SIZE_SPM, PAGE_SIZE);

    // let the VPEManager know about us before we continue with initialization
    VPEManager::get().add(this);

    // we have one reference to ourself
    rem_ref();

    if(~_flags & F_IDLE)
        init();

    KLOG(VPES, "Created VPE '" << _name << "' [id=" << id << ", pe=" << pe() << "]");
    for(auto &r : _requires)
        KLOG(VPES, "  requires: '" << r.name << "'");
}

VPE::~VPE() {
    KLOG(VPES, "Deleting VPE '" << _name << "' [id=" << id() << "]");

    _state = DEAD;

    if(_group)
        _group->remove(this);

    free_reqs();

    _objcaps.revoke_all();
    _mapcaps.revoke_all();

    // ensure that there are no syscalls for this VPE anymore
    DTU::get().drop_msgs(syscall_ep(), reinterpret_cast<label_t>(this));
    SyscallHandler::free_ep(syscall_ep());

    if(_rbufcpy)
        MainMemory::get().free(_rbufcpy);

    delete _as;

    VPEManager::get().remove(this);
}

void VPE::flush_cache() {
    if(_flags & F_FLUSHED)
        return;

    KLOG(VPES, "Flushing cache of VPE '" << _name << "' [id=" << id() << "]");

    VPE *cur = PEManager::get().current(pe());
    assert(cur != nullptr);
    DTU::get().flush_cache(cur->desc());
    _flags |= F_FLUSHED;
}

void VPE::make_daemon() {
    _flags |= F_DAEMON;
    VPEManager::get()._daemons++;
}

void VPE::start_app(int pid) {
    if(has_app())
        return;

    _pid = pid;
    _flags |= F_HASAPP;

    // when exiting, the program will release one reference
    add_ref();

    KLOG(VPES, "Starting VPE '" << _name << "' [id=" << id() << "]");

    PEManager::get().start_vpe(this);
}

void VPE::stop_app(int exitcode, bool self) {
    if(!has_app())
        return;

    if(self)
        exit_app(exitcode);
    else if(_state == RUNNING)
        exit_app(1);
    else {
        PEManager::get().stop_vpe(this);
        _flags ^= F_HASAPP;
    }

    if(rem_ref())
        delete this;
}

static int exit_event = 0;

void VPE::wait_for_exit() {
    m3::ThreadManager::get().wait_for(reinterpret_cast<event_t>(&exit_event));
}

void VPE::exit_app(int exitcode) {
    // no update on the PE here, since we don't save the state anyway
    _dtustate.invalidate_eps(m3::DTU::FIRST_FREE_EP);

    // "deactivate" send and receive gates
    for(capsel_t sel = 2; sel < 2 + EP_COUNT - m3::DTU::FIRST_FREE_EP; ++sel) {
        auto epcap = static_cast<EPCapability*>(_objcaps.get(sel, Capability::EP));
        if(epcap == nullptr || epcap->obj->gate == nullptr)
            continue;

        if(epcap->obj->gate->type == Capability::SGATE)
            static_cast<SGateObject*>(epcap->obj->gate)->activated = false;
        else if(epcap->obj->gate->type == Capability::RGATE)
            static_cast<RGateObject*>(epcap->obj->gate)->addr = 0;

        // forget the connection
        epcap->obj->gate->remove_ep(&*epcap->obj);
        epcap->obj->gate = nullptr;
    }

    _exitcode = exitcode;

    _flags ^= F_HASAPP;

    PEManager::get().stop_vpe(this);

    m3::ThreadManager::get().notify(reinterpret_cast<event_t>(&exit_event));
}

void VPE::yield() {
    if(_pending_fwds == 0)
        PEManager::get().yield_vpe(this);
}

bool VPE::migrate(bool fast) {
    // idle VPEs are never migrated
    if(_flags & (VPE::F_IDLE | VPE::F_PINNED))
        return false;

    peid_t old = pe();

    bool changed = PEManager::get().migrate_vpe(this, fast);
    if(changed)
        KLOG(VPES, "Migrated VPE '" << _name << "' [id=" << id() << "] from " << old << " to " << pe());

    return changed;
}

bool VPE::migrate_for(VPE *vpe) {
    if(_flags & (VPE::F_IDLE | VPE::F_PINNED))
        return false;

    peid_t old = pe();

    bool changed = PEManager::get().migrate_for(this, vpe);
    if(changed)
        KLOG(VPES, "Migrated VPE '" << _name << "' [id=" << id() << "] from " << old << " to " << pe());
    return changed;
}

bool VPE::resume(bool need_app, bool unblock) {
    if(need_app && !has_app())
        return false;

    KLOG(VPES, "Resuming VPE '" << _name << "' (unblock=" << unblock << ") [id=" << id() << "]");

    bool wait = true;
    if(unblock)
        wait = !PEManager::get().unblock_vpe(this, false);
    if(wait)
        m3::ThreadManager::get().wait_for(reinterpret_cast<event_t>(this));

    KLOG(VPES, "Resumed VPE '" << _name << "' [id=" << id() << "]");
    return true;
}

void VPE::wakeup() {
    if(_state == RUNNING)
        DTU::get().inject_irq(desc());
    else if(has_app())
        PEManager::get().unblock_vpe(this, false);
}

void VPE::notify_resume() {
    m3::ThreadManager::get().notify(reinterpret_cast<event_t>(this));
}

void VPE::free_reqs() {
    for(auto it = _requires.begin(); it != _requires.end(); ) {
        auto old = it++;
        delete &*old;
    }
}

void VPE::upcall_notify(m3::Errors::Code res, word_t event) {
    m3::KIF::Upcall::Notify msg;
    msg.opcode = m3::KIF::Upcall::NOTIFY;
    msg.error = res;
    msg.event = event;
    KLOG(UPCALLS, "Sending upcall NOTIFY (error=" << res << ", event="
        << (void*)event << ") to VPE " << id());
    upcall(&msg, sizeof(msg), false);
}

bool VPE::invalidate_ep(epid_t ep, bool cmd) {
    KLOG(EPS, "VPE" << id() << ":EP" << ep << " = invalid");

    bool res = true;
    if(cmd) {
        if(is_on_pe())
            res = DTU::get().inval_ep_remote(desc(), ep) == m3::Errors::NONE;
        else
            res = _dtustate.invalidate(ep, true);
    }
    else {
        _dtustate.invalidate(ep, false);
        update_ep(ep);
    }
    return res;
}

bool VPE::can_forward_msg(epid_t ep) {
    if(is_on_pe())
        _dtustate.read_ep(desc(), ep);
    return _dtustate.can_forward_msg(ep);
}

void VPE::forward_msg(epid_t ep, peid_t pe, vpeid_t vpe) {
    KLOG(EPS, "VPE" << id() << ":EP" << ep << " forward message");

    _dtustate.forward_msg(ep, pe, vpe);
    update_ep(ep);
}

void VPE::forward_mem(epid_t ep, peid_t pe) {
    KLOG(EPS, "VPE" << id() << ":EP" << ep << " forward mem");

    _dtustate.forward_mem(ep, pe);
    update_ep(ep);
}

m3::Errors::Code VPE::config_rcv_ep(epid_t ep, RGateObject &obj) {
    // it needs to be in the receive buffer space
    const goff_t addr = Platform::def_recvbuf(pe());
    const size_t size = Platform::pe(pe()).has_virtmem() ? RECVBUF_SIZE : RECVBUF_SIZE_SPM;
    // def_recvbuf() == 0 means that we do not validate it
    if(addr && (obj.addr < addr || obj.addr > addr + size || obj.addr + obj.size() > addr + size))
        return m3::Errors::INV_ARGS;
    if(obj.addr < addr + _rbufs_size)
        return m3::Errors::INV_ARGS;

    // no free headers left?
    size_t msgSlots = 1UL << (obj.order - obj.msgorder);
    if(_headers + msgSlots > m3::DTU::HEADER_COUNT)
        return m3::Errors::OUT_OF_MEM;

    obj.header = _headers;
    KLOG(EPS, "VPE" << id() << ":EP" << ep << " = "
        "RGate[addr=#" << m3::fmt(obj.addr, "x")
        << ", order=" << obj.order
        << ", msgorder=" << obj.msgorder
        << ", header=" << obj.header << "]");

    _dtustate.config_recv(ep, obj.addr, obj.order, obj.msgorder, obj.header);
    update_ep(ep);

    // TODO really manage the header space and zero the headers first in case they are reused
    _headers += msgSlots;

    m3::ThreadManager::get().notify(reinterpret_cast<event_t>(&obj));
    return m3::Errors::NONE;
}

m3::Errors::Code VPE::config_snd_ep(epid_t ep, SGateObject &obj) {
    assert(obj.rgate->addr != 0);
    if(obj.activated)
        return m3::Errors::EXISTS;

    peid_t pe = VPEManager::get().peof(obj.rgate->vpe);
    KLOG(EPS, "VPE" << id() << ":EP" << ep << " = "
        "Send[vpe=" << obj.rgate->vpe
        << ", pe=" << pe
        << ", ep=" << obj.rgate->ep
        << ", label=#" << m3::fmt(obj.label, "x")
        << ", msgsize=" << obj.rgate->msgorder << ", crd=" << obj.credits << "]");

    obj.activated = true;
    _dtustate.config_send(ep, obj.label, pe, obj.rgate->vpe,
                          obj.rgate->ep, 1UL << obj.rgate->msgorder, obj.credits);
    update_ep(ep);
    return m3::Errors::NONE;
}

m3::Errors::Code VPE::config_mem_ep(epid_t ep, const MGateObject &obj, goff_t off) {
    if(off >= obj.size || obj.addr + off < off)
        return m3::Errors::INV_ARGS;

    KLOG(EPS, "VPE" << id() << ":EP" << ep << " = "
        "Mem [vpe=" << obj.vpe
        << ", pe=" << obj.pe
        << ", addr=#" << m3::fmt(obj.addr + off, "x")
        << ", size=#" << m3::fmt(obj.size - off, "x")
        << ", perms=#" << m3::fmt(obj.perms, "x") << "]");

    // TODO
    _dtustate.config_mem(ep, obj.pe, obj.vpe, obj.addr + off, obj.size - off, obj.perms);
    update_ep(ep);
    return m3::Errors::NONE;
}

void VPE::update_ep(epid_t ep) {
    if(is_on_pe())
        DTU::get().write_ep_remote(desc(), ep, _dtustate.get_ep(ep));
}

}

File: src/apps/rust/kernel/src/pes/vpe.rs
use base::col::{String, ToString, Vec};
use base::cell::{Ref, RefCell, RefMut};
use base::cfg;
use base::dtu::{EpId, PEId, HEADER_COUNT, EP_COUNT, FIRST_FREE_EP};
use base::errors::{Code, Error};
use base::kif::{CapRngDesc, CapType, PEDesc, Perm};
use base::rc::Rc;
use core::fmt;

use arch::kdtu;
use cap::{Capability, CapTable, KObject, SGateObject, RGateObject, MGateObject};
use platform;

pub type VPEId = usize;

bitflags! {
    pub struct VPEFlags : u32 {
        const BOOTMOD     = 0b00000001;
        const DAEMON      = 0b00000010;
        const IDLE        = 0b00000100;
        const INIT        = 0b00001000;
        const HASAPP      = 0b00010000;
        const MUXABLE     = 0b00100000; // TODO temporary
        const READY       = 0b01000000;
        const WAITING     = 0b10000000;
    }
}

#[derive(Clone, Copy, Debug, Eq, PartialEq)]
pub enum State {
    RUNNING,
    SUSPENDED,
    DEAD
}

pub const INVALID_VPE: VPEId = 0xFFFF;

pub struct VPEDesc<'v> {
    pe: PEId,
    vpe: Option<&'v VPE>,
}

impl<'v> VPEDesc<'v> {
    pub fn new(pe: PEId, vpe: &'v VPE) -> VPEDesc<'v> {
        VPEDesc {
            pe: pe,
            vpe: Some(vpe),
        }
    }

    pub fn new_mem(pe: PEId) -> Self {
        VPEDesc {
            pe: pe,
            vpe: None,
        }
    }

    pub fn vpe(&self) -> Option<&VPE> {
        self.vpe
    }
    pub fn pe_id(&self) -> PEId {
        self.pe
    }
    pub fn vpe_id(&self) -> VPEId {
        match self.vpe {
            Some(v) => v.id(),
            None    => INVALID_VPE,
        }
    }
}

pub struct VPE {
    id: VPEId,
    pe: PEId,
    pid: i32,
    state: State,
    name: String,
    flags: VPEFlags,
    obj_caps: CapTable,
    eps_addr: usize,
    args: Vec<String>,
    req: Vec<String>,
    ep_caps: Vec<Option<KObject>>,
    dtu_state: kdtu::State,
    rbufs_size: usize,
    headers: usize,
}

impl VPE {
    pub fn new(name: &str, id: VPEId, pe: PEId, flags: VPEFlags) -> Rc<RefCell<Self>> {
        let vpe = Rc::new(RefCell::new(VPE {
            id: id,
            pe: pe,
            pid: 0,
            state: State::DEAD,
            name: name.to_string(),
            flags: flags,
            obj_caps: CapTable::new(),
            eps_addr: 0,
            args: Vec::new(),
            req: Vec::new(),
            ep_caps: vec![None; EP_COUNT - FIRST_FREE_EP],
            dtu_state: kdtu::State::new(),
            rbufs_size: 0,
            headers: 0,
        }));

        {
            let mut vpe_mut = vpe.borrow_mut();
            unsafe {
                vpe_mut.obj_caps.set_vpe(vpe.as_ptr());
            }

            // cap for own VPE
            vpe_mut.obj_caps_mut().insert(
                Capability::new(0, KObject::VPE(vpe.clone()))
            );
            // cap for own memory
            vpe_mut.obj_caps_mut().insert(
                Capability::new(1, KObject::MGate(MGateObject::new(
                    pe, id, 0, cfg::MEM_CAP_END, Perm::RWX
                )))
            );

            vpe_mut.init();
        }
        vpe
    }

    pub fn destroy(&mut self) {
        // remove the circular reference
        self.obj_caps.revoke(CapRngDesc::new(CapType::OBJECT, 0, 1), true);
    }

    #[cfg(target_os = "linux")]
    fn init(&mut self) {
    }

    #[cfg(target_os = "none")]
    fn init(&mut self) {
        use base::dtu;
        use base::cfg;
        use pes::vpemng;

        let rgate = RGateObject::new(cfg::SYSC_RBUF_ORD, cfg::SYSC_RBUF_ORD);

        // attach syscall receive endpoint
        {
            let mut rgate = rgate.borrow_mut();
            rgate.vpe = vpemng::KERNEL_VPE;
            rgate.order = cfg::SYSC_RBUF_ORD;
            rgate.msg_order = cfg::SYSC_RBUF_ORD;
            rgate.addr = platform::default_rcvbuf(self.pe_id());
            rgate.ep = Some(dtu::SYSC_SEP);
        }
        self.config_rcv_ep(dtu::SYSC_REP, &mut rgate.borrow_mut()).unwrap();

        // attach syscall endpoint
        let sgate = SGateObject::new(&rgate, self.id() as dtu::Label, cfg::SYSC_RBUF_SIZE as u64);
        self.config_snd_ep(dtu::SYSC_SEP, &sgate.borrow());

        // attach upcall receive endpoint
        {
            let mut rgate = rgate.borrow_mut();
            rgate.order = cfg::UPCALL_RBUF_ORD;
            rgate.msg_order = cfg::UPCALL_RBUF_ORD;
            rgate.addr += cfg::SYSC_RBUF_SIZE;
        }
        self.config_rcv_ep(dtu::UPCALL_REP, &mut rgate.borrow_mut()).unwrap();

        // attach default receive endpoint
        {
            let mut rgate = rgate.borrow_mut();
            rgate.order = cfg::DEF_RBUF_ORD;
            rgate.msg_order = cfg::DEF_RBUF_ORD;
            rgate.addr += cfg::DEF_RBUF_SIZE;
        }
        self.config_rcv_ep(dtu::DEF_REP, &mut rgate.borrow_mut()).unwrap();

        self.rbufs_size = rgate.borrow().addr + (1 << rgate.borrow().order);
        self.rbufs_size -= platform::default_rcvbuf(self.pe_id());
    }

    pub fn id(&self) -> VPEId {
        self.id
    }
    pub fn pe_id(&self) -> PEId {
        self.pe
    }
    pub fn desc(&self) -> VPEDesc {
        VPEDesc::new(self.pe, self)
    }
    pub fn pe_desc(&self) -> PEDesc {
        platform::pe_desc(self.pe_id())
    }
    pub fn name(&self) -> &str {
        &self.name
    }
    pub fn args(&self) -> &Vec<String> {
        &self.args
    }
    pub fn requirements(&self) -> &Vec<String> {
        &self.req
    }

    pub fn obj_caps(&self) -> &CapTable {
        &self.obj_caps
    }
    pub fn obj_caps_mut(&mut self) -> &mut CapTable {
        &mut self.obj_caps
    }

    pub fn state(&self) -> State {
        self.state.clone()
    }
    pub fn set_state(&mut self, state: State) {
        self.state = state;
    }

    pub fn dtu_state(&mut self) -> &mut kdtu::State {
        &mut self.dtu_state
    }

    pub fn make_daemon(&mut self) {
        self.flags |= VPEFlags::DAEMON;
    }
    pub fn add_arg(&mut self, arg: &str) {
        self.args.push(arg.to_string());
    }
    pub fn add_requirement(&mut self, req: &str) {
        self.req.push(req.to_string());
    }

    pub fn eps_addr(&self) -> usize {
        self.eps_addr
    }
    pub fn set_eps_addr(&mut self, eps: usize) {
        self.eps_addr = eps;
    }

    pub fn set_pid(&mut self, pid: i32) {
        self.pid = pid;
    }

    pub fn get_ep_cap(&self, ep: EpId) -> Option<KObject> {
        self.ep_caps[ep].clone()
    }
    pub fn set_ep_cap(&mut self, ep: EpId, cap: Option<KObject>) {
        self.ep_caps[ep] = cap;
    }

    pub fn config_snd_ep(&mut self, ep: EpId, obj: &Ref<SGateObject>) {
        let rgate = obj.rgate.borrow();
        assert!(rgate.activated());

        klog!(EPS, "VPE{}:EP{} = {:?}", self.id(), ep, obj);

        let pe_id = vpemng::get().pe_of(rgate.vpe).unwrap();
        self.dtu_state.config_send(
            ep, obj.label, pe_id, rgate.vpe, rgate.ep.unwrap(), rgate.msg_size(), obj.credits
        );
        self.update_ep(ep);
    }

    pub fn config_rcv_ep(&mut self, ep: EpId, obj: &mut RefMut<RGateObject>) -> Result<(), Error> {
        // it needs to be in the receive buffer space
        let addr = platform::default_rcvbuf(self.pe_id());
        let size = platform::rcvbufs_size(self.pe_id());

        // default_rcvbuf() == 0 means that we do not validate it
        if addr != 0 && (obj.addr < addr ||
                         obj.addr > addr + size ||
                         obj.addr + obj.size() > addr + size ||
                         obj.addr < addr + self.rbufs_size) {
            return Err(Error::new(Code::InvArgs));
        }

        // no free headers left?
        let msg_slots = 1 << (obj.order - obj.msg_order);
        if self.headers + msg_slots > HEADER_COUNT {
            return Err(Error::new(Code::OutOfMem));
        }

        // TODO really manage the header space and zero the headers first in case they are reused
        obj.header = self.headers;
        self.headers += msg_slots;

        klog!(EPS, "VPE{}:EP{} = {:?}", self.id(), ep, obj);

        self.dtu_state.config_recv(ep, obj.addr, obj.order, obj.msg_order, obj.header);
        self.update_ep(ep);

        Ok(())
    }

    pub fn config_mem_ep(&mut self, ep: EpId, obj: &Ref<MGateObject>, off: usize) -> Result<(), Error> {
        if off >= obj.size || obj.addr + off < off {
            return Err(Error::new(Code::InvArgs));
        }

        klog!(EPS, "VPE{}:EP{} = {:?}", self.id(), ep, obj);

        // TODO
        self.dtu_state.config_mem(ep, obj.pe, obj.vpe, obj.addr + off, obj.size - off, obj.perms);
        self.update_ep(ep);
        Ok(())
    }

    pub fn invalidate_ep(&mut self, ep: EpId, cmd: bool) -> Result<(), Error> {
        klog!(EPS, "VPE{}:EP{} = invalid", self.id(), ep);

        if cmd {
            if self.state == State::RUNNING {
                kdtu::KDTU::get().invalidate_ep_remote(&self.desc(), ep)?;
            }
            else {
                self.dtu_state.invalidate(ep, true)?;
            }
        }
        else {
            self.dtu_state.invalidate(ep, false)?;
            self.update_ep(ep);
        }
        Ok(())
    }

    fn update_ep(&mut self, ep: EpId) {
        if self.state == State::RUNNING {
            kdtu::KDTU::get().write_ep_remote(&self.desc(), ep, self.dtu_state.get_ep(ep)).unwrap();
        }
    }
}

impl fmt::Debug for VPE {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        write!(f, "VPE[id={}, pe={}, name={}, state={:?}]",
            self.id(), self.pe_id(), self.name(), self.state())
    }
}
